library/include/BitArray.h

```diff
@@ -232,4 +234,299 @@ namespace DFHack
+            {
+                return **this;
+            }
+
+            operator const_iterator() const
+            {
+                return const_iterator(*this);
+            }
+            bool operator==(const iterator & other) const
+            {
+                if (root != other.root)
+                    return false;
+
+                if (other.next && !next)
+                    return prev && other.prev && prev->next == other.prev;
+                if (next && !other.next)
+                    return prev && other.prev && other.prev->next == prev;
+                return other.prev == prev;
+            }
+            bool operator!=(const iterator & other) const
+            {
+                return !(*this == other);
+            }
+        };
+        class const_iterator
+        {
+            iterator iter;
+            friend struct DfLinkedList<L, I>;
+        public:
+            using difference_type = void;
+            using value_type = I * ;
+            using pointer = I * const *;
+            using reference = const proxy;
+            using iterator_category = std::bidirectional_iterator_tag;
+
+            const_iterator(const iterator & iter) : iter(iter) {}
+            const_iterator(const const_iterator & other) : iter(other.iter) {}
+
+            const_iterator & operator++()
+            {
+                ++iter;
+                return *this;
+            }
+            const_iterator & operator--()
+            {
+                --iter;
+                return *this;
+            }
+            const_iterator operator++(int)
+            {
+                const_iterator copy(*this);
+                ++*this;
+                return copy;
+            }
+            const_iterator operator--(int)
+            {
+                const_iterator copy(*this);
+                --*this;
+                return copy;
+            }
+            const_iterator & operator=(const const_iterator & other)
+            {
+                iter = other.iter;
+                return *this;
+            }
+            const proxy operator*() const
+            {
+                return *iter;
+            }
+            I *operator->() const
+            {
+                return *iter;
+            }
+            bool operator==(const const_iterator & other) const
+            {
+                return iter == other.iter;
+            }
+            bool operator!=(const const_iterator & other) const
+            {
+                return iter != other.iter;
+            }
+        };
+
+        typedef I *value_type;
+        typedef I * & reference_type;
+        typedef void difference_type;
+        typedef size_t size_type;
+
+        bool empty() const
+        {
+            return static_cast<const L *>(this)->next == nullptr;
+        }
+        size_t size() const
+        {
+            size_t n = 0;
+            for (I *i : *this)
+                n++;
+            return n;
+        }
```

Comment:
maybe return std::distance(begin(),end());

Reply:
You can't subtract the iterators in constant time because it's a linked list.